File: Books/A_Comprehensive_Approach_to_ML_and_AI/matlab/cnn_tutorial.m

close all; clear all; clc; 

%% Load and Preprocess Image
% For demonstration, we use the built-in 'cameraman.tif' image.
I = imread('data/cameraman.tif');
I = im2double(I); % Normalize image to [0,1]

% Display the original image
figure;
subplot(1,3,1);
imshow(I);
title('Original Image');

%% CNN Parameters
% Convolution kernel (e.g., edge detector)
K = [-1 -1 -1; 0 0 0; 1 1 1];
% Pooling size (2x2)
poolSize = 2;
% Fully connected layer parameters (random initialization)
% The size of the flattened feature map will be determined later.
% For demonstration, we set a single output neuron.

%% Forward Pass through CNN
% 1. Convolution: valid convolution
S = myConv2(I, K);

% 2. ReLU Activation
A = max(0, S);

% 3. Max Pooling
P = maxPool(A, poolSize);

% 4. Flatten the pooled feature map
f = P(:);

% 5. Fully connected layer: for demonstration, use random weights and bias.
W_fc = randn(1, numel(f)) * 0.01;
b_fc = 0;
y = W_fc * f + b_fc;

%% Visualization of Intermediate Steps
subplot(1,3,2);
    imagesc(A);
    colormap gray;
    title('ReLU Activation Output');
    axis image off;
subplot(1,3,3);
    imagesc(P);
    colormap gray;
    title('Max Pooling Output');
    axis image off;

save_all_figs_OPTION('cnn2','png',1); 

figure;
    subplot(1,2,1);
    imshow(I);
    title('Original Image');
subplot(1,2,2);
    imagesc(S);
    axis square
    colormap gray;
    title('Convolved Image (Valid Mode)');
    colorbar;    

% save_all_figs_OPTION('results/cnn3','png',1)

fprintf('The fully connected layer output (prediction) is: %.4f\n', y);


function S = myConv2(I, K)
    % MYCONV2 performs 2D convolution of matrix I with kernel K in valid mode.
    %   I: Input image (2D matrix)
    %   K: Convolution kernel (2D matrix)
    %   S: Output matrix after valid convolution
    %
    % The function computes the convolution result S by sliding the kernel K
    % over the input I. Only positions where the kernel fully overlaps with I
    % are considered ("valid" convolution).
    
    % Get dimensions of the input and the kernel
    [H, W] = size(I);
    [kH, kW] = size(K);
    
    % Calculate the dimensions of the output matrix
    outH = H - kH + 1;
    outW = W - kW + 1;
    
    % Pre-allocate the output matrix S with zeros
    S = zeros(outH, outW);
    
    % Loop over every valid position in I
    for i = 1:outH
        for j = 1:outW
            % Extract the current patch from I
            patch = I(i:i+kH-1, j:j+kW-1);
            % Compute the sum of elementwise multiplication of patch and K
            S(i,j) = sum(sum(patch .* K));
        end
    end
end


function P = maxPool(A, poolSize)
    % MAXPOOL performs max pooling on matrix A with a pooling window of size poolSize.
    [H, W] = size(A);
    Hp = floor(H / poolSize);
    Wp = floor(W / poolSize);
    P = zeros(Hp, Wp);
    for i = 1:Hp
        for j = 1:Wp
            patch = A((i-1)*poolSize+1:i*poolSize, (j-1)*poolSize+1:j*poolSize);
            P(i,j) = max(patch(:));
        end
    end
end

function s = sigmoid(x)
    % SIGMOID computes the sigmoid function.
    s = 1 ./ (1 + exp(-x));
end

File: Books/A_Comprehensive_Approach_to_ML_and_AI/python/cnn_tutorial.py
#!/usr/bin/env python3
"""
cnn_tutorial.py
---------------
This tutorial demonstrates a simple CNN on the “cameraman” image.
It applies a convolution (using a fixed kernel), ReLU activation, max pooling,
and a fully connected layer.
"""

import numpy as np
import matplotlib.pyplot as plt
from skimage import data, img_as_float

def my_conv2(I, K):
    H, W = I.shape
    kH, kW = K.shape
    outH = H - kH + 1
    outW = W - kW + 1
    S = np.zeros((outH, outW))
    for i in range(outH):
        for j in range(outW):
            patch = I[i:i+kH, j:j+kW]
            S[i,j] = np.sum(patch * K)
    return S

def max_pool(A, pool_size):
    H, W = A.shape
    Hp = H // pool_size
    Wp = W // pool_size
    P = np.zeros((Hp, Wp))
    for i in range(Hp):
        for j in range(Wp):
            patch = A[i*pool_size:(i+1)*pool_size, j*pool_size:(j+1)*pool_size]
            P[i,j] = np.max(patch)
    return P

# Load and preprocess image
I = img_as_float(data.camera())
plt.figure()
plt.subplot(1,3,1)
plt.imshow(I, cmap='gray')
plt.title("Original Image")
plt.axis('off')

# Convolution parameters: edge detector kernel
K = np.array([[-1, -1, -1],
              [ 0,  0,  0],
              [ 1,  1,  1]])
S = my_conv2(I, K)
A = np.maximum(0, S)  # ReLU
poolSize = 2
P = max_pool(A, poolSize)

# Fully connected layer (flatten pooled features)
f = P.flatten()
W_fc = 0.01 * np.random.randn(1, f.size)
b_fc = 0
y = W_fc @ f + b_fc

plt.subplot(1,3,2)
plt.imshow(A, cmap='gray')
plt.title("ReLU Activation Output")
plt.axis('off')
plt.subplot(1,3,3)
plt.imshow(P, cmap='gray')
plt.title("Max Pooling Output")
plt.axis('off')
plt.tight_layout()
plt.show()

plt.figure()
plt.subplot(1,2,1)
plt.imshow(I, cmap='gray')
plt.title("Original Image")
plt.axis('off')
plt.subplot(1,2,2)
plt.imshow(S, cmap='gray')
plt.title("Convolved Image (Valid)")
plt.colorbar()
plt.axis('off')
plt.tight_layout()
plt.show()

print(f"The fully connected layer output (prediction) is: {y[0]:.4f}")
